doc/examples/qsv_transcode.c

```diff
@@ -0,0 +1,431 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+/**
+ * @file
+ * Intel QSV-accelerated transcoding example.
+ *
+ * @example qsv_transcode.c
+ * This example shows how to do QSV-accelerated transcoding and how to
+ * dynamically change encoder's option.
+ * Usage: qsv_transcode input_stream codec output_stream initial option
+ *                      { frame_number new_option }
+ * e.g: - qsv_transcode input.mp4 h264_qsv output_h264.mp4 "g 60"
+ *      - qsv_transcode input.mp4 hevc_qsv output_hevc.mp4 "g 60 async_depth 1"
+ *                      100 "g 120"
+ *         (initialize codec with gop_size 60 and change it to 120 after 100
+ *          frames)
+ */
+
+#include <stdio.h>
+#include <errno.h>
+
+#include <libavutil/hwcontext.h>
+#include <libavcodec/avcodec.h>
+#include <libavformat/avformat.h>
+#include <libavutil/opt.h>
+
+static AVFormatContext *ifmt_ctx = NULL, *ofmt_ctx = NULL;
+static AVBufferRef *hw_device_ctx = NULL;
+static AVCodecContext *decoder_ctx = NULL, *encoder_ctx = NULL;
+static int video_stream = -1;
+static int initialized = 0;
+
+typedef struct DynamicSetting {
+    int frame_number;
+    char* optstr;
+} DynamicSetting;
+static DynamicSetting *dynamic_setting;
+static int setting_number;
+static int current_setting_number;
+
+static int str_to_dict(char* optstr, AVDictionary **opt)
+{
+    char *key, *value;
+    key = strtok(optstr, " ");
+    if (key == NULL)
+        return AVERROR(ENAVAIL);
+    value = strtok(NULL, " ");
+    if (value == NULL)
+        return AVERROR(ENAVAIL);
+    av_dict_set(opt, key, value, 0);
+    do {
+        key = strtok(NULL, " ");
+        if (key == NULL)
+            return 0;
+        value = strtok(NULL, " ");
+        if (value == NULL)
+            return AVERROR(ENAVAIL);
+        av_dict_set(opt, key, value, 0);
+    } while(key != NULL);
+    return 0;
+}
+
+static int dynamic_set_parameter(AVCodecContext *avctx)
+{
+    AVDictionary *opts = NULL;
+    int ret = 0;
+    static int frame_number = 0;
+    frame_number++;
+    if (current_setting_number < setting_number &&
+        frame_number == dynamic_setting[current_setting_number].frame_number) {
+        AVDictionaryEntry *e = NULL;
+        ret = str_to_dict(dynamic_setting[current_setting_number].optstr, &opts);
+        if (ret < 0) {
+            fprintf(stderr, "The dynamic parameter is wrong\n");
+            goto fail;
+        }
+        /* set comman option */
+        if ((ret = av_opt_set_dict(avctx, &opts)) < 0)
+            goto fail;
+        /* set codec specfic option */
+        if ((ret = av_opt_set_dict(avctx->priv_data, &opts)) < 0)
+            goto fail;
+        e = av_dict_get(opts, "r", NULL, 0);
```

Comment:
Can you explain what is this code on lines 103-107 about? Preferably as source comment.

Reply:
There is no "framerate" option, so I use "-r" to set framerate which is compatible with ffmpeg commandline. According to document of time_base: `For fixed-fps content, timebase should be 1/framerate and timestamp increments should be identically 1.` I set time_base as well. I have added comment to this part of code.

Reply:
I see, so if user tried to set a framerate, this will fail (with above av_opt_set_dict) calls and you need to do that explicitly working with context.

The problem of this method is that it's not atomic. I.e. potentially there might be a race when framerate will got updated and time_base will not.

Reply:
This example is single thread, so I think it is OK. If you want to make it pipeline, you may need to lock dynamic setting part.